streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java

```diff
@@ -185,12 +186,31 @@ private void cleanup(final long segmentId) {
         }
     }
 
-    private long segmentIdFromSegmentName(String segmentName) {
-        try {
-            Date date = formatter.parse(segmentName.substring(name.length() + 1));
-            return date.getTime() / segmentInterval;
-        } catch (Exception ex) {
-            return -1L;
+    private long segmentIdFromSegmentName(String segmentName, final File parent) {
+        // old style segment name with date
+        if (segmentName.charAt(name.length()) == '-') {
+            final String datePart = segmentName.substring(name.length() + 1);
+            final Date date;
+            try {
+                date = formatter.parse(datePart);
+            } catch (ParseException e) {
+                throw new ProcessorStateException("Unable to parse segmentName:"
+                                                          + segmentName
+                                                          + " to a date");
+            }
+            final long segmentId = date.getTime() / segmentInterval;
+            final File newName = new File(parent, segmentName(segmentId));
+            final File oldName = new File(parent, segmentName);
+            if (!oldName.renameTo(newName)) {
+                throw new ProcessorStateException("Unable to rename old style segment from: "
```

Comment:
If we don't throw exception here, future calls to segmentIdFromSegmentName would try to rename (and may succeed).
Is that better than throwing exception immediately ?

Reply:
There is no future call. This happens only when the database is being opened.

Reply:
Segments#getOrCreateSegment() would be called after the open.
If Segments class keeps a Set of segmentId's for which renameTo() was not successful, the old directory can still be used.
IMHO not being able to rename is not critical error. This should be rare case where the use of formatter has negligible impact on performance.

Reply:
It is not called from `Segments#getOrCreateSegment`

Reply:
segmentName() (where the format change happens) is called from Segments#getOrCreateSegment().

Reply:
File.renameTo can be unreliable on different platforms; dependent upon OS and FS. I've seen it fail erratically with no explanation and sometimes load related. The ugly fallback is to copy the file.

Reply:
@tedyu I think @dguy 's point is that `segmentIdFromSegmentName()` is only called when we are opening the database and trying to load the existing segments, where we would like to try rename the segment names if they are in the old format (as you mentioned earlier that `future calls to segmentIdFromSegmentName` actually would not happen).

I'm wondering, if it is really worthy renaming the files? Since the older segments would be removed eventually as time advances, and after we have opened the segment database we would only reference them by ids and not by segment names any more, maybe we do not need to rename them at all?

Reply:
That is correct @guozhangwang. As far not renaming the segment: there was a reason i did it, but i'll have to take a look at why that was the case.

Reply:
@guozhangwang, the segment needs to be renamed as the segment name is derived in `getOrCreateSegment` Line 86, i.e., it does `Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);`
at this point there is no way to tell if it was an old style name or a new style name.
Also, this renaming will only be a 1 off event, i.e., the first time it runs after an upgrade, so i feel it is fine to leave it as is.

Reply:
Sounds Good